044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,253 @@
+the cluster is becoming too full it will throttle production of messages. The kafka quota API isn't rich enough to do anything 
+smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
+
+### High level changes
+To better support external sources for managing quotas this proposal introduces some new concepts to the plugin:
+
+#### Quota Source
+
+We propose adding a Quota Source concept to the plugin to provide an extension point where we could plug in
+future sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
+
+We introduce the concept of throttle factor. This is a factor in the range (0.0, 1.0) that is applied to a quota to
+calculate the final byte rate limit. 
+
+This is a concept that was already implicit in the current calculations, but we want to name it so that we can use it in 
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
+
+#### Volume
+
+The storage quotas operate on observations about a **Volume** with these characteristics:
+1. `logDir`: the path of the logdir
+2. `brokerId`
+3. `capacityBytes`
+4. `usableBytes`
+
+#### Volume Source
+
+The quota plugin operates on observations about Volumes, the **Volume Source** determines where
+we obtain those observations. Values are:
+1. `local`: we inspect this brokers local log dirs to discover the state of their Volume
+2. `cluster`: we ask kafka for the state of all log dirs for all currently active nodes
+
+the source will be configurable with a property like:
+
+1. `client.quota.callback.static.storage.volume.source=cluster`
+
+Local will be the default if the property isn't provided but is deprecated from the beginning.
+
+#### Cluster Volume Source
+
+With the introduction of [KIP-827](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+logdirs+total+and+usable+space+via+Kafka+API) in
+kafka 3.3 we can now obtain the total and usable bytes per log dir as part of the DescribeLogDirsResponse.
+
+Note: if a single disk contains multiple log dirs, it will be described multiple times through the kafka APIs. This
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster.
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions though the admin api we need to construct an admin client. We intend to
+connect to the replication listener and will piggyback on the broker's existing SSL configuration to
+obtain keystore/truststore properties.
+
+If using cluster sourced volumes we will expose these optional properties to customise the admin connection:
+
+1. `client.quota.callback.kafka.listener.name` (default `replication-9091`)
+2. `client.quota.callback.kafka.listener.port` (default `9091`)
```

Comment:
port and protocol could be derived from the listener.name but we would then be parsing the values of the server's listener configuration properties to try and determine the port/protocol.

We thought it might be simpler to have them explicit rather than writing this logic to extract the details.

Reply:
Replaced with explicit configuration of admin client props